pkg/prober/prober.go

```diff
@@ -0,0 +1,155 @@
+package prober
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+	"sync"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/common/route"
+)
+
+const (
+	healthyEndpointPath  = "/-/healthy"
+	readyEndpointPath    = "/-/ready"
+	okProbeText          = "thanos %v is %v"
+	errorProbeText       = "thanos %v is not %v. Reason: %v"
+	probeErrorHTTPStatus = 500
+	initialErrorText     = "thanos %s is initializing"
+)
+
+// Prober represents health and readriness status of given compoent.
+type Prober struct {
+	logger       log.Logger
+	loggerMtx    sync.Mutex
+	componentMtx sync.Mutex
+	component    string
+	readyMtx     sync.Mutex
+	readiness    error
+	healthyMtx   sync.Mutex
+	healthiness  error
+}
+
+// SetLogger sets logger used by the Prober.
+func (p *Prober) SetLogger(logger log.Logger) {
+	p.loggerMtx.Lock()
+	defer p.loggerMtx.Unlock()
+	p.logger = logger
+}
+
+func (p *Prober) getLogger() log.Logger {
+	p.loggerMtx.Lock()
+	defer p.loggerMtx.Unlock()
+	return p.logger
+}
+
+// SetComponent sets component name of the Prober displayed in responses.
+func (p *Prober) SetComponent(component string) {
+	p.componentMtx.Lock()
+	defer p.componentMtx.Unlock()
+	p.component = component
+}
+
+func (p *Prober) getComponent() string {
+	p.componentMtx.Lock()
+	defer p.componentMtx.Unlock()
+	return p.component
+}
+
+// NewProber returns Prober reprezenting readiness and healthiness of given component.
+func NewProber(component string, logger log.Logger) *Prober {
+	initialErr := fmt.Errorf(initialErrorText, component)
+	prober := &Prober{}
+	prober.SetComponent(component)
+	prober.SetLogger(logger)
+	prober.SetNotHealthy(initialErr)
+	prober.SetNotReady(initialErr)
+	return prober
+}
+
+// HandleInMux registers readiness and liveness probes to mux.
+func (p *Prober) RegisterInRouter(router *route.Router) {
+	router.Get(healthyEndpointPath, p.probeHandlerFunc(p.IsHealthy, "healthy"))
+	router.Get(readyEndpointPath, p.probeHandlerFunc(p.IsReady, "ready"))
+}
+
+// NewProbeInRouter returns new Prober which registers it's ready and health endpoints to given router.
+func NewProbeInRouter(component string, router *route.Router, logger log.Logger) *Prober {
```

Comment:
Consider removing this, IMO 2 calls are better than having tons of different constructor methods.

Reply:
Yes it was almost unused. Refactoring lead to the 2 calls and this was the old way. removed, thanks